flocker/node/agents/test/test_blockdevice.py

```diff
@@ -2787,65 +2846,148 @@ class ResizeFilesystemInitTests(
+
+
 class ResizeFilesystemTests(
     make_istatechange_tests(
         ResizeFilesystem,
-        dict(volume=_ARBITRARY_VOLUME),
-        dict(volume=_ARBITRARY_VOLUME.set(blockdevice_id=u"wxyz")),
+        dict(volume=_ARBITRARY_VOLUME, size=REALISTIC_BLOCKDEVICE_SIZE),
+        dict(
+            volume=_ARBITRARY_VOLUME.set(blockdevice_id=u"wxyz"),
+            size=REALISTIC_BLOCKDEVICE_SIZE
+        ),
     ),
 ):
     """
     Tests for ``ResizeFilesystem``\ 's ``IStateChange`` implementation.
     """
-    def test_grow(self):
+    def test_size_invariant(self):
+        """
+        ``ResizeFilesystem.size`` must be a multiple of 512.
+        """
+        self.assertRaises(
+            InvariantException,
+            ResizeFilesystem,
+            volume=_ARBITRARY_VOLUME, size=513,
+        )
+
+    def _resize_test(self, size_factor):
         """
-        ``ResizeFilesystem.run`` increases the size of the filesystem on a
-        block device to the size of that block device.
+        Assert that ``ResizeFilesystem`` can change the size of an existing
+        dataset's filesystem.
+
+        :param size_factor: A multiplier to apply to the current size of the
+            filesystem to determine the new size that ``ResizeFilesystem`` will
+            be told to use.
+
+        :raise: A test-failing exception if ``ResizeFilesystem`` produces a
+            filesystem in which the number of inodes has not changed by a
+            factor of ``size_factor``.
         """
         host = u"192.0.7.8"
         dataset_id = uuid4()
 
+        filesystem = u"ext4"
+
+        original_size = REALISTIC_BLOCKDEVICE_SIZE
+        new_size = int(original_size * size_factor)
+
         deployer = create_blockdevicedeployer(self, hostname=host)
         api = deployer.block_device_api
-        mountpoint = deployer.mountroot.child(b"resized-filesystem")
 
         volume = api.create_volume(
-            dataset_id=dataset_id, size=REALISTIC_BLOCKDEVICE_SIZE,
+            dataset_id=dataset_id, size=original_size,
         )
-        filesystem = u"ext4"
-
-        attach = AttachVolume(dataset_id=dataset_id, hostname=host)
-        createfs = CreateFilesystem(volume=volume, filesystem=filesystem)
-        mount = MountBlockDevice(dataset_id=dataset_id, mountpoint=mountpoint)
+        api.attach_volume(volume.blockdevice_id, host)
 
-        unmount = UnmountBlockDevice(dataset_id=dataset_id)
-        detach = DetachVolume(dataset_id=dataset_id)
-        resize = ResizeVolume(
-            volume=volume, size=REALISTIC_BLOCKDEVICE_SIZE * 2
+        self.successResultOf(run_state_change(
+            CreateFilesystem(
+                volume=volume, filesystem=filesystem,
+            ),
+            deployer
+        ))
+
+        if new_size > original_size:
+            api.detach_volume(volume.blockdevice_id)
+            api.resize_volume(volume.blockdevice_id, new_size)
+            api.attach_volume(volume.blockdevice_id, host)
+
```

Comment:
`ResizeBlockDeviceDataset` detaches before resizing filesystem when shrinking, so this doesn't quite match actual interaction pattern.